File: cpp/cpp_100-999/623.cc
// C++
// leetcode 623
// https://leetcode.cn/problems/add-one-row-to-tree/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* addOneRow(TreeNode* root, int val, int depth) {
    	if (depth == 1) {
    		return new TreeNode(val, root, nullptr);
    	}

    	dfs(root, 1, depth, val);

    	return root;

    }

    void dfs(TreeNode *p, int cur, int target_depth, int val) {
    	if (cur >= target_depth)
    		return;

    	if (cur + 1 == target_depth) {
    		TreeNode *newL = new TreeNode(val, p -> left, nullptr);
    		TreeNode *newR = new TreeNode(val, nullptr, p -> right);
    		p -> left = newL;
    		p -> right = newR;
    		return ;
    	}

    	if (p -> left)
    		dfs(p -> left, cur + 1, target_depth, val);

    	if (p -> right)
    		dfs(p -> right, cur + 1, target_depth, val);
    }
};

File: Go/Go_100-999/623.go
// Go
// leetcode 623
// https://leetcode.cn/problems/add-one-row-to-tree/

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func addOneRow(root *TreeNode, val int, depth int) *TreeNode {
	if depth == 1 {
		return &TreeNode{val, root, nil}
	}

	dfs(root, 1, val, depth)

	return root
}

func dfs(p *TreeNode, cur, val, depth int) {
	if cur >= depth {
		return
	}

	if cur+1 == depth {
		p.Left = &TreeNode{val, p.Left, nil}
		p.Right = &TreeNode{val, nil, p.Right}
		return
	}

	if p.Left != nil {
		dfs(p.Left, cur+1, val, depth)
	}

	if p.Right != nil {
		dfs(p.Right, cur+1, val, depth)
	}
}